### PR TITLE
Issue 674 replace starlette middleware types

### PR DESCRIPTION
**Pull request by Goldziher**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [ ] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

### CODE REVIEW

File: starlite/middleware/authentication.py [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 from abc import ABC, abstractmethod
-from typing import TYPE_CHECKING, Any, List, Optional, Pattern, Union
+from typing import TYPE_CHECKING, Any, List, Optional, Pattern, Set, Union
 
 from pydantic import BaseConfig, BaseModel
 
 from starlite.connection import ASGIConnection
 from starlite.enums import ScopeType
-from starlite.middleware.util import should_bypass_middleware
+from starlite.middleware.utils import should_bypass_middleware
```
Comment (jtraub) [inline starlite/middleware/authentication.py:9]: Maybe we should update CSRF middleware as well in this PR. It also relies on exclude opts to skip protection on certain routes.

I think @provinzkraut simply forgot to uppdate it in #630 when he introduced should_bypass_middleware function

Comment (Goldziher) [reply starlite/middleware/authentication.py:9]: good point, but it should be done in a followup

Comment (provinzkraut) [reply starlite/middleware/authentication.py:9]: Yup, I did!

File: starlite/middleware/base.py [hunk 2]
```diff
@@ -65,3 +82,91 @@ def __call__(self, app: "ASGIApp") -> "ASGIApp":
...
         return self.middleware(*self.args, app=app, **self.kwargs)
+
+
+class _AbstractMiddlewareMetaClass(ABCMeta):
+    app: "ASGIApp"
+    exclude: Optional[Union[str, List[str]]]
+    exclude_opt_key: Optional[str]
+    scopes: Set["Literal[ScopeType.HTTP, ScopeType.WEBSOCKET]"]
+
+    def __new__(cls, name: str, bases: Tuple[Type, ...], namespace: Dict[str, Callable], **kwargs: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply starlite/middleware/base.py:93]: Pylint complains, there is a rule specifically against this.

[the remark this answers is not in the document]
Comment (provinzkraut) [reply starlite/middleware/base.py:93]: `cls` is indeed the canonical name here: https://docs.python.org/3/reference/datamodel.html#object.__new__

**Review by provinzkraut — APPROVED** [review]

**Pull request closed without merging**